examples/step-69/helper.h

```diff
@@ -0,0 +1,83 @@
+#ifndef HELPER_H
+#define HELPER_H
+
+#include <deal.II/base/function.h>
+#include <deal.II/base/tensor.h>
+
+#include <type_traits>
+
+/*
+ * A convenience macro that automatically writes out an accessor (or
+ * getter) function:
+ *
+ *   const Foo& bar() const
+ *   {
+ *      return bar_;
+ *   }
+ *
+ * or
+ *
+ *   const Foo& bar() const
+ *   {
+ *      return *bar_;
+ *   }
+ *
+ * depending on whether bar_ can be dereferenced, or not.
+ */
+
+namespace
+{
+  template <typename T>
+  class is_dereferenciable
+  {
+    template <typename C>
+    static auto test(...) -> std::false_type;
+
+    template <typename C>
+    static auto test(C *) -> decltype(*std::declval<C>(), std::true_type());
+
+  public:
+    typedef decltype(test<T>(nullptr)) type;
+    static constexpr auto              value = type::value;
+  };
+
+  template <typename T, typename>
+  auto dereference(T &t) -> decltype(dereference(*t)) &;
+
+  template <
+    typename T,
+    typename = typename std::enable_if<!is_dereferenciable<T>::value>::type>
+  const T &dereference(T &t)
+  {
+    return t;
+  }
+
+  template <
+    typename T,
+    typename = typename std::enable_if<is_dereferenciable<T>::value>::type>
+  auto dereference(T &t) -> const decltype(*t) &
+  {
+    return *t;
+  }
+} /* anonymous namespace */
+
+#define ACCESSOR_READ_ONLY(member)                 \
+public:                                            \
```

Comment:
ah, the macro is here. My preference would be to just get rid of it and spell things out -- that's what we do everywhere else.

Reply:
@bangerth None of this stuff made it into the last iteration of the code. Sorry for having had a very old version pushed temporarily :-/